### PR TITLE
[AIRFLOW-6175] fixes scheduler queue bug

### PR DESCRIPTION
**Pull request by dimberman**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6175
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

There is a bug caused by scheduler_jobs refactor which leads to task failure and scheduler locking.

Essentially when a there is an overflow of tasks going into the scheduler, the tasks are set back to scheduled, but are not removed from the executor's queued_tasks queue.

This means that the executor will attempt to run tasks that are in the scheduled state, but those tasks will fail dependency checks. Eventually the queue is filled with scheduled tasks, and the scheduler can no longer run.
### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by ashb** [comment]
Need a jira for this please @dimberman

File: tests/executors/test_executor.py [hunk 1]
```diff
@@ -55,9 +55,11 @@ def sort_by(item):
                 (_, prio, _, _) = val
                 # Sort by priority (DESC), then date,task, try
                 return -prio, date, dag_id, task_id, try_number
-            sorted_queue = sorted(self.queued_tasks.items(), key=sort_by)
 
-            for (key, (_, _, _, simple_ti)) in sorted_queue:
+            open_slots = self.parallelism - len(self.running)
+            sorted_queue = sorted(self.queued_tasks.items(), key=sort_by)
+            for _ in range(min((open_slots, len(sorted_queue)))):
+                (key, (_, _, _, simple_ti)) = sorted_queue.pop()
```
Comment (ashb) [inline tests/executors/test_executor.py]: This change is just to  make the test executor behave more like the other ones that have limited parallelism right?

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -188,6 +188,55 @@ def test_process_executor_events(self, mock_stats_incr):
 
         mock_stats_incr.assert_called_once_with('scheduler.tasks.killed_externally')
 
+    def test_scheduler_executor_overflow(self):
+        """
+        Test that tasks that tasks that are set back to scheduled after
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]: after what? :)

File: tests/jobs/test_scheduler_job.py [hunk 3]
```diff
@@ -188,6 +188,55 @@ def test_process_executor_events(self, mock_stats_incr):
...
+        """
+        executor = TestExecutor(do_update=True, parallelism=3)
+        session = settings.Session()
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]: Using `with create_session() as session:` is what we do elsewhere - could you do that for consistency please?

File: tests/jobs/test_scheduler_job.py [hunk 4]
```diff
@@ -188,6 +188,55 @@ def test_process_executor_events(self, mock_stats_incr):
...
+            tis.append(ti)
+            session.merge(ti)
+        session.commit()
+
+        simple_dag_bag = self._make_simple_dag_bag([dag])
+        scheduler = SchedulerJob(num_runs=1,
+                                 executor=executor,
+                                 subdir=os.path.join(settings.DAGS_FOLDER,
+                                                     "no_dags.py"))
+        mock.patch.object(scheduler, '_change_state_for_tis_without_dagrun').start()
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]: Do we need to patch this?

File: tests/jobs/test_scheduler_job.py [hunk 5]
```diff
@@ -188,6 +188,55 @@ def test_process_executor_events(self, mock_stats_incr):
...
+        scheduler._process_dags(simple_dag_bag)
+        [ti.refresh_from_db() for ti in tis]
+        self.assertEqual(len(executor.queued_tasks), 0)
+        successful_tasks = [ti for ti in tis if ti.state == State.SUCCESS]
+        scheduled_tasks = [ti for ti in tis if ti.state == State.SCHEDULED]
+        self.assertEqual(3, len(successful_tasks))
+        self.assertEqual(6, len(scheduled_tasks))
+
+    def create_test_dag(self):
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]: Could you use this in the other places in the file too? (Will probably need to make it look like this:

```suggestion
    def create_test_dag(self, start_date=DEFAULT_DATE, **kwargs):
    dag = DAG(start_date=start_date, **kwargs)
```

File: tests/jobs/test_scheduler_job.py (hunk 3 above)
Comment (ashb) [reply tests/jobs/test_scheduler_job.py]: (oh, though we don't to that in this test file yet. Still, we should for the new code we add)

**Comment by dimberman** [comment]
@ashb this bug has a jira https://issues.apache.org/jira/browse/AIRFLOW-6175

File: tests/executors/test_executor.py (hunk 1 above)
Comment (dimberman) [reply tests/executors/test_executor.py]: Yes. We can't test overflowing if the executor "runs" all tasks so we had to create a bottleneck :).

**Comment by kaxil** [comment]
We have merge conflicts  :( - can you fix that please

File: tests/jobs/test_scheduler_job.py (hunk 4 above)
Comment (dimberman) [reply tests/jobs/test_scheduler_job.py]: Yes. Otherwise the tasks won't run because they don't have an associated DagRun.

**Comment by dimberman** [comment]
@kaxil fixed conflicts. @ashb addressed comments.

**Review by KevinYang21 — APPROVED** [review]
Thanks for fixing this! I'm wondering why the executor starts to run task in the scheduled state. Thought any task sent to the executor would be removed from queue_tasks and be moved to running_tasks and thus the state won't be set back to scheduled.

File: tests/jobs/test_scheduler_job.py [hunk 6]
```diff
@@ -188,6 +204,41 @@ def test_process_executor_events(self, mock_stats_incr):
 
         mock_stats_incr.assert_called_once_with('scheduler.tasks.killed_externally')
 
+    def test_scheduler_executor_overflow(self):
+        """
+        Test that tasks that tasks that are set back to scheduled and removed from the executor
```
Comment (kaxil) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        Test that tasks that are set back to scheduled and removed from the executor
```

File: tests/jobs/test_scheduler_job.py [hunk 7]
```diff
@@ -188,6 +204,41 @@ def test_process_executor_events(self, mock_stats_incr):
...
+        queue in the case of an overflow.
+        """
+        executor = MockExecutor(do_update=True, parallelism=3)
+        with create_session() as session:
+            dag = self.create_test_dag()
+            task = DummyOperator(
+                task_id='dummy',
+                dag=dag,
+                owner='airflow')
+            from datetime import timedelta
```
Comment (kaxil) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
```

Already defined on L25

File: tests/jobs/test_scheduler_job.py [hunk 8]
```diff
@@ -188,6 +204,41 @@ def test_process_executor_events(self, mock_stats_incr):
...
+            tis = []
+            for i in range(1, 10):
+                ti = TI(task, DEFAULT_DATE + timedelta(days=i))
+                ti.state = State.SCHEDULED
+                tis.append(ti)
+                session.merge(ti)
+            session.commit()
```
Comment (kaxil) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
```

`create_session` would commit the session when exiting the contextmanager

File: tests/jobs/test_scheduler_job.py [hunk 9]
```diff
@@ -188,6 +204,41 @@ def test_process_executor_events(self, mock_stats_incr):
...
+
+            simple_dag_bag = self._make_simple_dag_bag([dag])
+            scheduler = SchedulerJob(num_runs=1,
+                                     executor=executor,
+                                     subdir=os.path.join(settings.DAGS_FOLDER,
+                                                         "no_dags.py"))
+            mock.patch.object(scheduler, '_change_state_for_tis_without_dagrun').start()
+            scheduler._process_dags(simple_dag_bag)
+            [ti.refresh_from_db() for ti in tis]
```
Comment (kaxil) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
            for ti in tis:
                ti.refresh_from_db() 
```

as we don't need to create list

**Review by kaxil — APPROVED** [review]
Minor suggestion, otherwise LGTM

**Comment by KevinYang21** [comment]
Let me answer my own question... Since we change from SCHEDULED to QUEUED only when the dag is in the simple_dagbag, thus those left over tasks in queued_tasks won't be set back to QUEUED until the next parse but stlll gonna be picked up by the executor, then cause the deps check to fail.

Thanks again for fixing my bad 🙏 

And to understand the impact completely, the bug was causing scheduler/executor to slow down quite a lot when traffic >> executor capacity, but not completely blocked right? As those SCHEDULED tasks in queued_tasks would be slowly moved out by _execute_task_instances.

**Comment by kaxil** [comment]
Following tests are failing:
```
FAILED tests/jobs/test_backfill_job.py::TestBackfillJob::test_backfill_examples_0_example_branch_operator
FAILED tests/jobs/test_backfill_job.py::TestBackfillJob::test_backfill_examples_1_example_bash_operator
FAILED tests/jobs/test_backfill_job.py::TestBackfillJob::test_backfill_examples_2_example_skip_dag
FAILED tests/jobs/test_backfill_job.py::TestBackfillJob::test_backfill_multi_dates
FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_retry_still_in_executor
FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_scheduler_executor_overflow
FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_scheduler_reschedule
==== 7 failed, 3999 passed, 117 skipped, 16 warnings in 1481.02s (0:24:41) =====
```

**Review by potiuk — APPROVED** [review]
Nice one! Let's get it merged. We have some other "global" updates to this code which aim to increase readability (pylint + type annotations) so we should -  I think - merge those fairly quickly.

**Comment by dimberman** [comment]
@potiuk @KevinYang21 wanna give a final lookover before merging?

**Comment by KevinYang21** [comment]
oops, just found a debug line that I forgot to remove https://github.com/apache/airflow/pull/6732/commits/c0ea272a5cf6de16830bb7777b2ed9cc922821d1#diff-d7e6d11bfaeca087149ad577dd7704daR68. Otherwise LGTM. Will push a small commit to fix it. If we all good I can merge it after the CI passes.

**Review by mik-laj — APPROVED** [review]

**Pull request merged**